src/main/java/com/mongodb/jdbc/oidc/OidcAuthFlow.java

```diff
@@ -0,0 +1,159 @@
+        if (clientID == null || clientID.isEmpty()) {
+            logger.severe("Client ID is null or empty");
+            return null;
+        }
+
+        String issuerURI = idpServerInfo.getIssuer();
+        if (!issuerURI.startsWith("https")) {
+            logger.severe("Issuer URI must be HTTPS");
+            return null;
+        }
+
+        List<String> scopesList = idpServerInfo.getRequestScopes();
+        Scope scopes = new Scope();
+        if (scopesList != null) {
+            for (String scope : scopesList) {
+                scopes.add(new Scope.Value(scope));
+            }
+        }
+        // mongodb is not configured to ask for offline_access by default. We prefer always getting a
+        // refresh token when the server allows it.
+        scopes.add(new Scope.Value("offline_access"));
+
+        RFC8252HttpServer server = new RFC8252HttpServer();
+        try {
+            // Resolve OIDC provider metadata using the issuer URI and
+            // extract authorization and token endpoint URIs.
+            OIDCProviderMetadata providerMetadata =
+                    OIDCProviderMetadata.resolve(new Issuer(issuerURI));
+            URI authorizationEndpoint = providerMetadata.getAuthorizationEndpointURI();
+            URI tokenEndpoint = providerMetadata.getTokenEndpointURI();
+
+            // Start the local RFC8252 HTTP server to receive the redirect.
+            server.start();
+
+            URI redirectURI =
+                    new URI(
+                            "http://localhost:"
+                                    + RFC8252HttpServer.DEFAULT_REDIRECT_PORT
+                                    + "/redirect");
+            State state = new State();
+            CodeVerifier codeVerifier = new CodeVerifier();
+
+            // Build the authorization request URI.
+            AuthorizationRequest request =
+                    new AuthorizationRequest.Builder(
+                                    new ResponseType(ResponseType.Value.CODE),
+                                    new ClientID(clientID))
+                            .scope(scopes)
+                            .redirectionURI(redirectURI)
+                            .state(state)
+                            .codeChallenge(codeVerifier, CodeChallengeMethod.S256)
+                            .endpointURI(authorizationEndpoint)
+                            .build();
+
+            // Open the browser to the authorization request URI.
+            if (Desktop.isDesktopSupported()) {
+                Desktop.getDesktop().browse(request.toURI());
+            } else {
+                logger.severe("Desktop operations not supported");
+                return null;
+            }
+
+            // Wait for the authorization response from the local HTTP server.
+            OidcResponse response = server.getOidcResponse();
+            if (response == null || !state.getValue().equals(response.getState())) {
```

Comment:
To be clear, this is the client side (us) confirming that no man-in-the-middle attack occurred. This is a very important step. The code verifier is how the IdP ensures no man-in-the-middle attack occurred, since it could happen in either direction